Drop StringImpl::characters8() / characters16()

cdumez opened the pull request:
#### 67a879c31de358001ebd9ab1006d534c14bd8de5
<pre>
Drop StringImpl::characters8() / characters16()
<a href="https://bugs.webkit.org/show_bug.cgi?id=273376">https://bugs.webkit.org/show_bug.cgi?id=273376</a>

Reviewed by Darin Adler.

Drop StringImpl::characters8() / characters16() in favor of span equivalents.

* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::CachedUniquedStringImplBase::encode):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::FastStringifier&lt;CharType&gt;::append):
* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::JSString::dumpToStream):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanConstCast):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::domainToASCII):
* Source/WTF/wtf/text/AtomString.cpp:
(WTF::AtomString::convertASCIICase const):
* Source/WTF/wtf/text/AtomStringImpl.cpp:
(WTF::HashAndUTF8CharactersTranslator::equal):
* Source/WTF/wtf/text/StringBuilder.cpp:
(WTF::StringBuilder::shrink):
(WTF::StringBuilder::extendBufferForAppendingWithUpconvert):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::append):
(WTF::StringBuilder::characters const):
* Source/WTF/wtf/text/StringBuilderInternals.h:
(WTF::StringBuilder::extendBufferForAppending):
(WTF::StringBuilder::extendBufferForAppendingSlowCase):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::simplifyMatchedCharactersToSpace):
(WTF::StringImpl::find):
(WTF::equalInner):
(WTF::equalInternal):
(WTF::equal):
(WTF::equalIgnoringNullity):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::span8 const):
(WTF::StringImpl::span16 const):
(WTF::StringImpl::span&lt;LChar&gt; const):
(WTF::StringImpl::span&lt;UChar&gt; const):
(WTF::StringImpl::containsOnlyLatin1 const):
(WTF::StringImpl::removeCharactersImpl):
(WTF::StringImpl::removeCharacters):
(WTF::StringImpl::characters8 const): Deleted.
(WTF::StringImpl::characters16 const): Deleted.
(WTF::StringImpl::span const): Deleted.
(WTF::StringImpl::characters&lt;LChar&gt; const): Deleted.
(WTF::StringImpl::characters&lt;UChar&gt; const): Deleted.
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::characterAt const):
* Source/WTF/wtf/text/cf/StringImplCF.cpp:
(WTF::StringImpl::createCFString):
* Source/WTF/wtf/unicode/UTF8Conversion.cpp:
(WTF::Unicode::equalUTF16WithUTF8):
(WTF::Unicode::equalLatin1WithUTF8):
* Source/WTF/wtf/unicode/UTF8Conversion.h:
* Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
(WebCore::CSSTokenizerInputStream::advanceUntilNonWhitespace):
* Source/WebCore/css/parser/CSSTokenizerInputStream.h:
(WebCore::CSSTokenizerInputStream::characterPredicate):
* Source/WebCore/dom/DatasetDOMStringMap.cpp:
(WebCore::convertPropertyNameToAttributeName):
* Source/WebCore/platform/text/SegmentedString.h:
(WebCore::SegmentedString::Substring::Substring):

Canonical link: <a href="https://commits.webkit.org/278108@main">https://commits.webkit.org/278108@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a5c2e98b788621d607b215a7188d9a6ff644f69

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49476 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28758 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52514 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52715 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/149 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51780 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34776 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26378 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40389 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51576 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26301 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42626 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21507 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23761 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43809 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7845 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/42788 "Found 6 new JSC stress test failures: microbenchmarks/super-get-by-val-with-this-polymorphic.js.ftl-eager-no-cjit, stress/proxy-get-set-correct-receiver.js.ftl-eager-no-cjit, stress/proxy-has-property.js.ftl-eager-no-cjit, stress/proxy-own-keys.js.ftl-eager-no-cjit, stress/typedarray-defineOwnProperty-error.js.ftl-eager-no-cjit, stress/v8-typedarray-resizablearraybuffer.js.ftl-eager-no-cjit (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45675 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44315 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54227 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/48969 "Found 3 new JSC stress test failures: stress/poly-proto-using-inheritance.js.ftl-eager-no-cjit, stress/proxy-get-missing-handler-inline-cache.js.ftl-eager-no-cjit, stress/super-get-by-id.js.ftl-eager-no-cjit (failure)") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24559 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20743 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47758 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25830 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42815 "Build is in progress. Recent messages:") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46774 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26670 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/56459 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7115 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25553 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11599 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -849,7 +850,7 @@ size_t StringImpl::find(std::span<const LChar> matchString, size_t start)
 
         size_t i = 0;
-        while (searchHash != matchHash || !equal(searchCharacters + i, matchString)) {
+        while (searchHash != matchHash || !equal(searchCharacters.data() + i, matchString)) {
```
Kinda wish this was `subspace` instead, but I suppose that would require changing `equal`. Applies to other cases below.

cdumez replied:
Yes, I agree. It is a bigger change though so I'll address separately.

cdumez committed:
Drop StringImpl::characters8() / characters16()
https://bugs.webkit.org/show_bug.cgi?id=273376

Reviewed by Darin Adler.

Drop StringImpl::characters8() / characters16() in favor of span equivalents.

* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::CachedUniquedStringImplBase::encode):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::FastStringifier<CharType>::append):
* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::JSString::dumpToStream):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanConstCast):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::domainToASCII):
* Source/WTF/wtf/text/AtomString.cpp:
(WTF::AtomString::convertASCIICase const):
* Source/WTF/wtf/text/AtomStringImpl.cpp:
(WTF::HashAndUTF8CharactersTranslator::equal):
* Source/WTF/wtf/text/StringBuilder.cpp:
(WTF::StringBuilder::shrink):
(WTF::StringBuilder::extendBufferForAppendingWithUpconvert):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::append):
(WTF::StringBuilder::characters const):
* Source/WTF/wtf/text/StringBuilderInternals.h:
(WTF::StringBuilder::extendBufferForAppending):
(WTF::StringBuilder::extendBufferForAppendingSlowCase):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::simplifyMatchedCharactersToSpace):
(WTF::StringImpl::find):
(WTF::equalInner):
(WTF::equalInternal):
(WTF::equal):
(WTF::equalIgnoringNullity):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::span8 const):
(WTF::StringImpl::span16 const):
(WTF::StringImpl::span<LChar> const):
(WTF::StringImpl::span<UChar> const):
(WTF::StringImpl::containsOnlyLatin1 const):
(WTF::StringImpl::removeCharactersImpl):
(WTF::StringImpl::removeCharacters):
(WTF::StringImpl::characters8 const): Deleted.
(WTF::StringImpl::characters16 const): Deleted.
(WTF::StringImpl::span const): Deleted.
(WTF::StringImpl::characters<LChar> const): Deleted.
(WTF::StringImpl::characters<UChar> const): Deleted.
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::characterAt const):
* Source/WTF/wtf/text/cf/StringImplCF.cpp:
(WTF::StringImpl::createCFString):
* Source/WTF/wtf/unicode/UTF8Conversion.cpp:
(WTF::Unicode::equalUTF16WithUTF8):
(WTF::Unicode::equalLatin1WithUTF8):
* Source/WTF/wtf/unicode/UTF8Conversion.h:
* Source/WebCore/css/parser/CSSTokenizerInputStream.cpp:
(WebCore::CSSTokenizerInputStream::advanceUntilNonWhitespace):
* Source/WebCore/css/parser/CSSTokenizerInputStream.h:
(WebCore::CSSTokenizerInputStream::characterPredicate):
* Source/WebCore/dom/DatasetDOMStringMap.cpp:
(WebCore::convertPropertyNameToAttributeName):
* Source/WebCore/platform/text/SegmentedString.h:
(WebCore::SegmentedString::Substring::Substring):

Canonical link: https://commits.webkit.org/278108@main
